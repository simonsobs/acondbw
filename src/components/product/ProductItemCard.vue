<template>
  <v-card
    outlined
    hover
    class="product-item-card"
    style="max-width: 980px; position: relative"
  >
    <v-card-text v-if="loading">
      <v-progress-circular
        indeterminate
        :size="18"
        :width="3"
        color="grey"
      ></v-progress-circular>
    </v-card-text>
    <v-alert v-else-if="error" outlined dense type="error" class="ma-2">{{
      error
    }}</v-alert>
    <v-container
      v-else-if="loaded"
      @click="$emit('expand')"
      style="cursor: default"
      fluid
      class=""
    >
      <v-row>
        <v-col order="1" cols="9" md="4">
          <div class="caption grey--text">Name</div>
          <div class="font-weight-bold primary--text">
            <span @click.stop>
              <router-link
                :to="{
                  name: 'ProductItem',
                  params: {
                    productTypeName: node.type_.name,
                    name: node.name,
                  },
                }"
                v-text="node.name"
              ></router-link>
            </span>
          </div>
        </v-col>
        <v-col order="3" cols="6" md="3">
          <div class="caption grey--text">Date produced</div>
          <div v-text="node.dateProduced"></div>
        </v-col>
        <v-col order="4" cols="6" md="2">
          <div class="caption grey--text">Produced by</div>
          <div v-text="node.producedBy"></div>
        </v-col>
        <v-col order="2" order-md="5" cols="3" align-self="center">
          <v-container fluid>
            <v-row justify="end">
              <v-col v-if="collapsible" style="flex: 0" class="pa-0">
                <v-tooltip bottom open-delay="800">
                  <template v-slot:activator="{ on }">
                    <v-btn
                      icon
                      @click.stop="
                        collapsed ? $emit('expand') : $emit('collapse')
                      "
                      v-on="on"
                    >
                      <v-icon>
                        {{ collapsed ? "mdi-chevron-down" : "mdi-chevron-up" }}
                      </v-icon>
                    </v-btn>
                  </template>
                  <span>{{ collapsed ? "Expand" : "Collapse" }}</span>
                </v-tooltip>
              </v-col>
              <v-col @click.stop style="flex: 0" class="pa-0">
                <v-menu
                  left
                  bottom
                  offset-y
                  v-model="menu"
                  :close-on-content-click="false"
                >
                  <template v-slot:activator="{ on: menu }">
                    <v-btn icon v-on="{ ...menu }">
                      <v-icon>mdi-dots-vertical</v-icon>
                    </v-btn>
                  </template>
                  <v-list dense>
                    <v-dialog
                      v-model="editDialog"
                      persistent
                      fullscreen
                      transition="dialog-bottom-transition"
                    >
                      <template v-slot:activator="{ on: editDialog }">
                        <v-list-item
                          :disabled="disableEdit"
                          v-on="{ ...editDialog }"
                        >
                          <v-list-item-icon>
                            <v-icon :disabled="disableEdit">mdi-pencil</v-icon>
                          </v-list-item-icon>
                          <v-list-item-content>
                            <v-list-item-title>Update</v-list-item-title>
                          </v-list-item-content>
                        </v-list-item>
                      </template>
                      <v-card>
                        <v-app-bar
                          dense
                          color="secondary"
                          style="position: sticky; top: 0; z-index: 999"
                        >
                          <v-btn
                            icon
                            dark
                            @click="
                              menu = false;
                              editDialog = false;
                            "
                          >
                            <v-icon>mdi-close</v-icon>
                          </v-btn>
                        </v-app-bar>
                        <ProductEditForm
                          :productId="node.productId"
                          v-on:finished="
                            menu = false;
                            editDialog = false;
                          "
                        ></ProductEditForm>
                      </v-card>
                    </v-dialog>
                    <v-dialog v-model="deleteDialog" max-width="600">
                      <template v-slot:activator="{ on: deleteDialog }">
                        <v-list-item
                          :disabled="disableDelete"
                          v-on="{ ...deleteDialog }"
                        >
                          <v-list-item-icon>
                            <v-icon :disabled="disableDelete"
                              >mdi-delete</v-icon
                            >
                          </v-list-item-icon>
                          <v-list-item-content>
                            <v-list-item-title>Delete</v-list-item-title>
                          </v-list-item-content>
                        </v-list-item>
                      </template>
                      <ProductDeleteForm
                        :productId="node.productId"
                        v-on:finished="
                          deleteDialog = false;
                          menu = false;
                        "
                        v-on:deleted="
                          deleteDialog = false;
                          menu = false;
                          node = null;
                          $emit('deleted');
                        "
                      ></ProductDeleteForm>
                    </v-dialog>
                  </v-list>
                </v-menu>
              </v-col>
            </v-row>
          </v-container>
        </v-col>
      </v-row>
      <v-expand-transition>
        <div class="collapsible" v-show="!(collapsible && collapsed)">
          <v-row>
            <v-col cols="12" md="4" offset-md="4">
              <div class="caption grey--text">Contact</div>
              <div v-text="node.contact"></div>
            </v-col>
          </v-row>
          <v-row>
            <v-col cols="12" md="8" offset-md="4">
              <div class="caption grey--text">Paths</div>
              <ul v-if="node.paths && node.paths.edges.length > 0">
                <li
                  v-for="(edgep, index) in node.paths.edges"
                  :key="index"
                  v-text="edgep.node.path"
                ></li>
              </ul>
              <div v-else class="body-2 grey--text">None</div>
            </v-col>
          </v-row>
          <v-row>
            <v-col cols="12" md="8" offset-md="4">
              <div class="caption grey--text">Relations</div>
              <div v-if="relations && Object.keys(relations).length > 0">
                <div v-for="(redges, typeId) in relations" :key="typeId">
                  <span class="capitalize subtitle-2 primary--text">
                    <span v-if="redges.length > 1">{{
                      redges[0].node.type_.plural
                    }}</span>
                    <span v-else>{{ redges[0].node.type_.singular }}</span
                    >:
                  </span>
                  <span v-for="(redge, index) in redges" :key="index">
                    <router-link
                      class="font-weight-bold primary--text"
                      :to="{
                        name: 'ProductItem',
                        params: {
                          productTypeName: redge.node.other.type_.name,
                          name: redge.node.other.name,
                        },
                      }"
                      v-text="redge.node.other.name"
                    ></router-link>
                    ({{ redge.node.other.type_.name }})<span
                      v-if="index != redges.length - 1"
                      >,
                    </span>
                  </span>
                </div>
              </div>
              <div v-else class="body-2 grey--text">None</div>
            </v-col>
          </v-row>
          <v-row>
            <v-col order="1" cols="12" md="8" offset-md="4">
              <div class="caption grey--text">Note</div>
              <div v-if="note" v-html="note"></div>
              <div v-else class="body-2 grey--text">None</div>
            </v-col>
          </v-row>
          <v-row
            no-gutters
            align="end"
            justify="space-between"
            class="grey--text mt-3"
            style="font-size: 80%"
          >
            <div>
              <div v-if="timeUpdated || node.updatingGitHubUser">
                Updated
                <span v-if="timeUpdated">at {{ timeUpdated }}</span>
                <span v-if="node.updatingGitHubUser">
                  by {{ node.updatingGitHubUser.login }}</span
                >
              </div>
              <div v-if="timePosted || node.postingGitHubUser">
                Posted
                <span v-if="timePosted">at {{ timePosted }}</span>
                <span v-if="node.postingGitHubUser">
                  by {{ node.postingGitHubUser.login }}</span
                >
              </div>
            </div>
            <div>Data ID: {{ dataId }}</div>
          </v-row>
        </div>
      </v-expand-transition>
    </v-container>
    <v-card-text v-else-if="notFound">Not Found</v-card-text>
    <dev-tool-loading-state-overriding-menu
      @state="devtoolState = $event"
    ></dev-tool-loading-state-overriding-menu>
  </v-card>
</template>

<script>
import _ from "lodash";

import marked from "marked";

import { defaultDataIdFromObject } from "apollo-cache-inmemory";

import PRODUCT from "@/graphql/queries/Product.gql";

import ProductEditForm from "@/components/product/ProductEditForm";
import ProductDeleteForm from "@/components/product/ProductDeleteForm";

import State from "@/utils/LoadingState.js";
import DevToolLoadingStateOverridingMenu from "@/components/utils/DevToolLoadingStateOverridingMenu";

export default {
  name: "ProductItemCard",
  components: {
    ProductEditForm,
    ProductDeleteForm,
    DevToolLoadingStateOverridingMenu,
  },
  props: {
    productId: { default: null }, // node.productId not node.id
    collapsed: { default: false },
    collapsible: { default: false },
    disableEdit: { default: false },
    disableDelete: { default: false },
  },
  data() {
    return {
      menu: false,
      editDialog: false,
      deleteDialog: false,
      init: true,
      node: null,
      error: null,
      devtoolState: null,
      State: State,
    };
  },
  computed: {
    state() {
      if (this.devtoolState) {
        return this.devtoolState;
      }

      if (this.$apollo.queries.node.loading) {
        return State.LOADING;
      } else if (this.error) {
        return State.ERROR;
      } else if (this.node) {
        return State.LOADED;
      } else if (this.init) {
        return State.INIT;
      } else {
        return State.NONE;
      }
    },
    loading() {
      return this.state == State.LOADING;
    },
    loaded() {
      return this.state == State.LOADED;
    },
    notFound() {
      return this.state == State.NONE;
    },
    dataId: function () {
      return defaultDataIdFromObject(this.node);
    },
    timePosted() {
      return this.node.timePosted
        ? this.formatDateTime(this.node.timePosted)
        : null;
    },
    timeUpdated() {
      return this.node.timeUpdated
        ? this.formatDateTime(this.node.timeUpdated)
        : null;
    },
    note() {
      return this.node.note ? marked(this.node.note) : null;
    },
    relations() {
      if (this.node && this.node.relations.edges.length > 0) {
        return _.groupBy(this.node.relations.edges, "node.type_.typeId");
      } else {
        return null;
      }
    },
  },
  watch: {
    devtoolState: function () {
      if (this.devtoolState) {
        this.init = this.devtoolState == State.INIT;
      }
      this.error =
        this.devtoolState == State.ERROR ? "Error from Dev Tools" : null;
    },
  },
  apollo: {
    node: {
      query: PRODUCT,
      variables() {
        return {
          productId: this.productId,
        };
      },
      update: (data) => data.product,
      result(result) {
        this.init = false;
        this.error = result.error ? result.error : null;
      },
    },
  },
  methods: {
    formatDateTime(dateTime) {
      const sinceEpoch = Date.parse(dateTime);
      const format = Intl.DateTimeFormat("default", {
        year: "numeric",
        month: "numeric",
        day: "numeric",
        hour: "numeric",
        minute: "numeric",
        second: "numeric",
        hour12: false,
      });
      return format.format(sinceEpoch);
    },
  },
};
</script>

<style scoped>
.capitalize {
  text-transform: capitalize;
}
</style>
