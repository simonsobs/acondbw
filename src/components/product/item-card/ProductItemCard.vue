<template>
  <v-card
    variant="outlined"
    hover
    class="product-item-card"
    style="position: relative"
  >
    <v-card-text v-if="loading">
      <v-progress-circular
        indeterminate
        :size="18"
        :width="3"
        color="secondary"
      >
      </v-progress-circular>
    </v-card-text>
    <v-alert v-else-if="error" variant="tonal" type="error" :text="error">
    </v-alert>
    <v-container
      v-else-if="loaded && node"
      @click="$emit('expand')"
      style="cursor: default"
    >
      <v-row>
        <v-col order="1" cols="9" md="4">
          <div class="text-h5 font-weight-medium text-primary">
            <span @click.stop>
              <router-link
                :to="{
                  name: 'ProductItem',
                  params: {
                    productTypeName: node.type_.name,
                    name: node.name,
                  },
                }"
                v-text="node.name"
                v-if="node.type_"
                class="text-decoration-none"
                style="color: inherit"
              >
              </router-link>
            </span>
          </div>
        </v-col>
        <v-col order="3" cols="6" md="3">
          <div class="text-caption">Date produced</div>
          <div
            v-if="attributes"
            v-text="attributes['date_produced']['value']"
          ></div>
        </v-col>
        <v-col order="4" cols="6" md="2">
          <div class="text-caption">Produced by</div>
          <div
            v-if="attributes"
            v-text="attributes['produced_by']['value']"
          ></div>
        </v-col>
        <v-col order="2" order-md="5" cols="3" align-self="center">
          <v-container>
            <v-row justify="end">
              <v-col v-if="collapsible" style="flex: 0" class="pa-0">
                <v-btn
                  v-bind="props"
                  variant="plain"
                  icon
                  @click.stop="collapsed ? $emit('expand') : $emit('collapse')"
                >
                  <v-icon
                    :icon="collapsed ? 'mdi-chevron-down' : 'mdi-chevron-up'"
                  >
                  </v-icon>
                </v-btn>
              </v-col>
              <v-col @click.stop style="flex: 0" class="pa-0">
                <v-menu
                  left
                  bottom
                  offset-y
                  v-model="menu"
                  :close-on-content-click="false"
                >
                  <template v-slot:activator="{ props }">
                    <v-btn v-bind="props" variant="plain" icon>
                      <v-icon icon="mdi-dots-vertical"></v-icon>
                    </v-btn>
                  </template>
                  <v-list dense>
                    <v-dialog v-model="editDialog" max-width="800" persistent>
                      <template v-slot:activator="{ props }">
                        <v-list-item
                          v-bind="props"
                          :disabled="disableEdit"
                          prepend-icon="mdi-pencil"
                          title="Edit"
                        >
                        </v-list-item>
                      </template>
                      <product-edit-form
                        v-if="editDialog && attributes"
                        :node="node"
                        :attributes="attributes"
                        @cancel="onEditFormCancelled"
                        @finished="onEditFormFinished($event)"
                      ></product-edit-form>
                    </v-dialog>
                    <v-dialog
                      v-model="updateRelationsDialog"
                      max-width="800"
                      persistent
                    >
                      <template v-slot:activator="{ props }">
                        <v-list-item
                          v-bind="props"
                          :disabled="disableEdit"
                          prepend-icon="mdi-relation-many-to-many"
                          title="Update relations"
                        >
                        </v-list-item>
                      </template>
                      <product-update-relations-form
                        v-if="updateRelationsDialog"
                        :node="node"
                        @cancel="onUpdateRelationsFormCancelled"
                        @finished="onUpdateRelationsFormFinished"
                      ></product-update-relations-form>
                    </v-dialog>
                    <v-dialog
                      v-model="convertTypeDialog"
                      max-width="800"
                      persistent
                    >
                      <template v-slot:activator="{ props }">
                        <v-list-item
                          v-bind="props"
                          :disabled="disableEdit"
                          prepend-icon="mdi-drawing"
                          title="Convert type"
                        >
                        </v-list-item>
                      </template>
                      <product-convert-type-form
                        v-if="convertTypeDialog"
                        :node="node"
                        @cancel="onConvertTypeFormCancelled"
                        @finished="onConvertTypeFormFinished($event)"
                      ></product-convert-type-form>
                    </v-dialog>
                    <v-dialog v-model="deleteDialog" max-width="600">
                      <template v-slot:activator="{ props }">
                        <v-list-item
                          v-bind="props"
                          :disabled="disableDelete"
                          prepend-icon="mdi-delete"
                          title="Delete"
                        >
                        </v-list-item>
                      </template>
                      <product-delete-form
                        v-if="deleteDialog"
                        :productId="Number(node.productId)"
                        @cancel="onDeleteFormCancelled"
                        @finished="onDeleteFormFinished"
                      ></product-delete-form>
                    </v-dialog>
                  </v-list>
                </v-menu>
              </v-col>
            </v-row>
          </v-container>
        </v-col>
      </v-row>
      <v-expand-transition>
        <div class="collapsible" v-show="!(collapsible && collapsed)">
          <v-row>
            <v-col cols="12" md="4" offset-md="4">
              <div class="text-caption">Contact</div>
              <div
                v-if="attributes"
                v-text="attributes['contact']['value']"
              ></div>
            </v-col>
          </v-row>
          <v-row>
            <v-col cols="12" md="8" offset-md="4">
              <div class="text-caption">Paths</div>
              <ul v-if="node.paths && node.paths.edges.length > 0">
                <li
                  v-for="(edgep, index) in node.paths.edges"
                  :key="index"
                  v-text="edgep && edgep.node && edgep.node.path"
                ></li>
              </ul>
              <div v-else class="text-body-2">None</div>
            </v-col>
          </v-row>
          <v-row>
            <v-col cols="12" md="8" offset-md="4">
              <div class="text-caption">Relations</div>
              <relations
                :relations="node.relations"
                v-if="node.relations"
              ></relations>
            </v-col>
          </v-row>
          <v-row>
            <v-col order="1" cols="12" md="8" offset-md="4">
              <div class="text-caption">Note</div>
              <note :markdown="node.note" v-if="node.note"></note>
              <div v-else class="text-body-2">None</div>
            </v-col>
          </v-row>
          <v-row
            no-gutters
            align="end"
            justify="space-between"
            class="mt-3"
            style="font-size: 11px"
          >
            <div>
              <div v-if="timeUpdated || node.updatingGitHubUser">
                Updated
                <span v-if="timeUpdated"> at {{ timeUpdated }} </span>
                <span v-if="node.updatingGitHubUser">
                  by {{ node.updatingGitHubUser.login }}
                </span>
              </div>
              <div v-if="timePosted || node.postingGitHubUser">
                Posted
                <span v-if="timePosted"> at {{ timePosted }} </span>
                <span v-if="node.postingGitHubUser">
                  by {{ node.postingGitHubUser.login }}
                </span>
              </div>
            </div>
            <div>ID: {{ node.id }}</div>
          </v-row>
        </div>
      </v-expand-transition>
    </v-container>
    <v-card-text v-else-if="notFound">Not Found</v-card-text>
    <dev-tool-loading-state-menu top="10px" right="10px" v-model="devtoolState">
    </dev-tool-loading-state-menu>
  </v-card>
</template>

<script setup lang="ts">
import { ref, computed, withDefaults } from "vue";

import ProductEditForm from "@/components/product/ProductEditForm.vue";
import ProductUpdateRelationsForm from "@/components/product/ProductUpdateRelationsForm.vue";
import ProductConvertTypeForm from "@/components/product/ProductConvertTypeForm.vue";
import ProductDeleteForm from "@/components/product/ProductDeleteForm.vue";

import { useProductQuery } from "@/generated/graphql";

import { useQueryState } from "@/utils/query-state";

import Relations from "./Relations.vue";
import Note from "./Note.vue";

interface Attribute {
  fieldId: number;
  name: string;
  value: unknown;
}

interface Attributes {
  [key: string]: Attribute;
}

const props = withDefaults(
  defineProps<{
    productId: number;
    collapsed?: boolean;
    collapsible?: boolean;
    disableEdit?: boolean;
    disableDelete?: boolean;
  }>(),
  {
    collapsed: false,
    collapsible: false,
    disableEdit: false,
    disableDelete: false,
  }
);

interface Emits {
  (e: "nameChanged", value: string | null): void;
  (e: "typeChanged", value: string | null): void;
  (e: "deleted", value: void): void;
  (e: "expand", value: void): void;
  (e: "collapse", value: void): void;
}

const emit = defineEmits<Emits>();

const query = useProductQuery({ variables: { productId: props.productId } });
const node = computed(() => query.data?.value?.product);
const timePosted = computed(() => formatDateTime(node.value?.timePosted));
const timeUpdated = computed(() => formatDateTime(node.value?.timeUpdated));

const attributes = computed<Attributes | null>(() => {
  if (!node.value) return null;
  const thisNode = node.value;

  const keys = [
    "attributesUnicodeText",
    "attributesBoolean",
    "attributesInteger",
    "attributesFloat",
    "attributesDate",
    "attributesDateTime",
    "attributesTime",
  ].filter((k) => k in thisNode);

  const ret = keys.reduce(
    (a, key) => ({
      ...a,
      ...thisNode[key].edges.reduce(
        (r, { node }) => ({
          ...r,
          ...{
            [node.field.name]: {
              fieldId: node.fieldId,
              name: node.field.name,
              value: node.value,
            },
          },
        }),
        {}
      ),
    }),
    {} as Attributes
  );
  return ret;
});
function formatDateTime(dateTime: string | undefined | null) {
  if (!dateTime) return null;
  const sinceEpoch = Date.parse(dateTime);
  const format = Intl.DateTimeFormat("default", {
    year: "numeric",
    month: "numeric",
    day: "numeric",
    hour: "numeric",
    minute: "numeric",
    second: "numeric",
    hour12: false,
  });
  return format.format(sinceEpoch);
}
const menu = ref(false);
const editDialog = ref(false);
function onEditFormCancelled() {
  closeEditForm();
}
function onEditFormFinished(event: string | undefined) {
  closeEditForm();
  if (event) emit("nameChanged", event);
}
function closeEditForm() {
  editDialog.value = false;
  menu.value = false;
}
const updateRelationsDialog = ref(false);
function onUpdateRelationsFormCancelled() {
  closeUpdateRelationsForm();
}
function onUpdateRelationsFormFinished() {
  closeUpdateRelationsForm();
}
function closeUpdateRelationsForm() {
  updateRelationsDialog.value = false;
  menu.value = false;
}
const convertTypeDialog = ref(false);
function onConvertTypeFormCancelled() {
  closeConvertTypeForm();
}
function onConvertTypeFormFinished(event: string) {
  closeConvertTypeForm();
  if (event) emit("typeChanged", event);
}
function closeConvertTypeForm() {
  convertTypeDialog.value = false;
  menu.value = false;
}
const deleteDialog = ref(false);
function onDeleteFormCancelled() {
  closeDeleteForm();
}
function onDeleteFormFinished() {
  closeDeleteForm();
  emit("deleted");
}
function closeDeleteForm() {
  deleteDialog.value = false;
  menu.value = false;
}

const queryState = useQueryState(query, { isNull: () => node.value === null });
const { loading, loaded, notFound, error, devtoolState } = queryState;
</script>
