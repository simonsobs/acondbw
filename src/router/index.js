import Vue from "vue";
import VueRouter from "vue-router";
import VueMeta from "vue-meta";

import Frame from "@/components/layout/Frame";
import NullFrame from "@/components/layout/NullFrame";

import Home from "@/views/framework/Home.vue";
import Entry from "@/views/framework/Entry.vue";

import OAuthRedirect from "@/views/auth/OAuthRedirect.vue"

import SignIn from "@/views/auth/SignIn.vue";
import Auth from "@/views/auth/Auth.vue";
import SignInError from "@/views/auth/SignInError.vue";

import ProductTop from "@/views/product/ProductTop.vue";
import ProductList from "@/views/product/ProductList.vue";
import ProductItem from "@/views/product/ProductItem.vue";
import ProductAdd from "@/views/product/ProductAdd.vue";

import store from "@/store";

Vue.use(VueRouter);
Vue.use(VueMeta);

const routes = [
  {
    path: "/",
    name: "Entry",
    components: {
      default: Entry,
      frame: NullFrame,
    },
    beforeEnter: (to, from, next) => {
      const signedIn = !!store.state.auth.token;
      if (signedIn) {
        next({ name: "Dashboard" });
      } else {
        next();
      }
    },
  },
  {
    path: "/dashboard",
    name: "Dashboard",
    components: {
      default: Home,
      frame: Frame,
    },
    meta: { requiresAuth: true },
    beforeEnter: (to, from, next) => {
      const signedIn = !!store.state.auth.token;
      if (signedIn) {
        next();
      } else {
        next({ name: "Entry" });
      }
    },
  },
  {
    path: "/oauth-redirect",
    name: "OAuthRedirect",
    components: {
      default: OAuthRedirect,
      frame: NullFrame,
    },
  },
  {
    path: "/signin",
    name: "SignIn",
    components: {
      default: SignIn,
      frame: NullFrame,
    },
  },
  {
    path: "/auth",
    name: "Auth",
    components: {
      default: Auth,
      frame: NullFrame,
    },
  },
  {
    path: "/signin-error",
    name: "SignInError",
    components: {
      default: SignInError,
      frame: NullFrame,
    },
  },
  {
    path: "/admin/scratch",
    name: "AdminScratch",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/Scratch"
        ),
      frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/admin/token",
    name: "AdminAppToken",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppToken"
        ),
      frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/admin/auth",
    name: "AdminAppAuth",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppAuth"
        ),
        frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/admin/token-error",
    name: "AdminAppTokenError",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppTokenError"
        ),
        frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/about",
    name: "About",
    component: () =>
      import(/* webpackChunkName: "about" */ "@/views/framework/About.vue"),
  },
  {
    path: "/add-product/:productTypeName",
    components: {
      default: ProductAdd,
      frame: NullFrame,
    },
    meta: { requiresAuth: true }
  },
  {
    path: "/product/:productTypeName",
    components: {
      default: ProductTop,
      frame: Frame,
    },
    meta: { requiresAuth: true },
    children: [
      {
        path: "",
        name: "ProductList",
        component: ProductList,
      },
      {
        path: ":name",
        name: "ProductItem",
        component: ProductItem,
      },
    ],
  },
  {
    path: "*",
    name: "NotFound",
    components: {
      default: () =>
        import(/* webpackChunkName: "error" */ "@/views/framework/NotFound.vue"),
        frame: NullFrame,
    },
  },
];

const router = new VueRouter({
  mode: "history",
  base: process.env.BASE_URL,
  routes,
});

router.beforeEach((to, from, next) => {
  const authRequired = to.matched.some((record) => record.meta.requiresAuth);
  const signedIn = !!store.state.auth.token;
  if (authRequired) {
    if (signedIn) {
      next();
    } else {
      next("/");
    }
  } else {
    next();
  }
});

export default router;
