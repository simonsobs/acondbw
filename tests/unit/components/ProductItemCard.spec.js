import Vue from "vue";
import VueRouter from "vue-router";
import Vuetify from "vuetify";
import { mount, createLocalVue } from "@vue/test-utils";

import ProductItemCard from "@/components/ProductItemCard.vue";
import router from "@/router";

Vue.use(Vuetify);
Vue.use(VueRouter);

describe("ProductItemCard.vue", () => {
  let localVue;
  let vuetify;

  function createWrapper({ loading = false, propsData } = {}) {
    const mutate = jest.fn();
    let wrapper = mount(ProductItemCard, {
      localVue,
      router,
      vuetify,
      mocks: {
        $apollo: {
          queries: {
            node: {
              loading: loading,
            },
          },
          mutate,
        },
      },
      propsData: {
        nameOfRouteToProductItem: "MapItem",
        productId: 1013,
        ...propsData,
      },
    });

    return wrapper;
  }

  const node = {
    id: "UHJvZHVjdDoxMDEz",
    productId: "1013",
    typeId: 1,
    type_: {
      id: "UHJvZHVjdFR5cGU6MQ==",
      typeId: "1",
      name: "map",
    },
    name: "lat20200201",
    contact: "pwg-pmn",
    dateProduced: "2020-02-01",
    producedBy: "pwg-pmn",
    datePosted: "2020-02-01",
    postedBy: "pwg-pmn",
    dateUpdated: "2020-02-01",
    updatedBy: "pwg-pmn",
    paths: {
      edges: [
        {
          node: {
            id: "UHJvZHVjdEZpbGVQYXRoOjQ=",
            pathId: "4",
            path: "nersc:/go/to/my/maps_v3",
            note: "lat only",
          },
        },
      ],
    },
    relations: {
      edges: [
        {
          node: {
            id: "UHJvZHVjdFJlbGF0aW9uOjQ=",
            relationId: "4",
            typeId: 2,
            type_: {
              id: "UHJvZHVjdFJlbGF0aW9uVHlwZToy",
              typeId: "2",
              name: "child",
            },
            otherProductId: 1150,
            other: {
              id: "UHJvZHVjdDoxMTUw",
              productId: "1150",
              typeId: 2,
              type_: {
                id: "UHJvZHVjdFR5cGU6Mg==",
                typeId: "2",
                name: "beam",
              },
              name: "20200207",
            },
            reverseRelationId: 3,
            reverse: {
              id: "UHJvZHVjdFJlbGF0aW9uOjM=",
              relationId: "3",
              typeId: 1,
              type_: {
                id: "UHJvZHVjdFJlbGF0aW9uVHlwZTox",
                typeId: "1",
                name: "parent",
              },
            },
          },
        },
      ],
    },
    note: "- This is a dummy test with a lat map\n- A beam depends on this map",
  };

  beforeEach(function() {
    localVue = createLocalVue();
    vuetify = new Vuetify();
  });

  it("loading", async () => {
    const loading = true;
    const wrapper = createWrapper({ loading });
    await Vue.nextTick();
    expect(wrapper.find(".v-progress-circular").exists()).toBe(true);
  });

  it("error", async () => {
    const wrapper = createWrapper();
    wrapper.setData({
      error: "Error: cannot load data",
    });
    await Vue.nextTick();
    expect(wrapper.text()).toContain("Error: cannot load data");
  });

  it("none", async () => {
    const wrapper = createWrapper();
    await Vue.nextTick();
    expect(wrapper.text()).toContain("Nothing to show here");
  });

  it("match snapshot", async () => {
    const wrapper = createWrapper();
    wrapper.setData({
      node: node,
    });
    await Vue.nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it.each([
    [true, true, false],
    [true, false, true],
    [false, true, true],
    [false, false, true],
  ])(
    "collapsed - {collapsible: %p, collapsed: %p, visible: %p}",
    async (collapsible, collapsed, visible) => {
      const wrapper = createWrapper({
        propsData: { collapsed: collapsed, collapsible: collapsible },
      });
      wrapper.setData({
        node: node,
      });
      await Vue.nextTick();
      expect(wrapper.find(".collapsible").isVisible()).toBe(visible);
    }
  );
});
