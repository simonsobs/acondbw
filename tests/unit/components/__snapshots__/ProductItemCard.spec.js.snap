// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductItemCard.vue match snapshot 1`] = `
<div class="product-item-card" style="position: relative;">
  <div class="v-card v-card--hover v-card--outlined v-sheet theme--light" style="max-width: 980px;">
    <div class="" style="cursor: default;">
      <div class="container pa-0 container--fluid">
        <div class="row ma-0 px-0">
          <div class="py-2 col-md-4 col-9 order-1">
            <div class="caption grey--text">Name</div>
            <div class="font-weight-bold primary--text"><span><a href="/map/item/lat20200201" class="">lat20200201</a></span></div>
          </div>
          <div class="py-2 col-md-3 col-6 order-3">
            <div class="caption grey--text">Date produced</div>
            <div>2020-02-01</div>
          </div>
          <div class="py-2 col-md-2 col-6 order-4">
            <div class="caption grey--text">Produced by</div>
            <div>pwg-pmn</div>
          </div>
          <div class="py-2 order-md-5 col col-3 order-2 align-self-center">
            <div class="row px-1 py-0 align-start justify-end">
              <!----> <span><button type="button" class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-dots-vertical theme--light"></i></span></button>
              <div class="v-menu">
                <!---->
              </div>
              </span>
            </div>
          </div>
        </div>
        <div class="row mx-0 px-0 collapsible">
          <div class="py-2 col-md-4 col-12">
            <div class="caption grey--text">Contact</div>
            <div>pwg-pmn</div>
          </div>
          <div class="py-2 col-md-3 col-6">
            <div class="caption grey--text">Date posted</div>
            <div>2020-02-01</div>
          </div>
          <div class="py-2 col-md-5 col-6">
            <div class="caption grey--text">Posted by</div>
            <div>pwg-pmn</div>
          </div>
          <div class="py-2 col-md-3 offset-md-4 col-6">
            <div class="caption grey--text">Date updated</div>
            <div>2020-02-01</div>
          </div>
          <div class="py-2 col-md-5 col-6">
            <div class="caption grey--text">Updated by</div>
            <div>pwg-pmn</div>
          </div>
          <div class="py-2 col-md-8 offset-md-4 col-12">
            <div class="caption grey--text">Paths</div>
            <ul>
              <li>nersc:/go/to/my/maps_v3</li>
            </ul>
          </div>
          <div class="py-2 col-md-8 offset-md-4 col-12">
            <div class="caption grey--text">Relations</div>
            <div>
              <div><span class="subtitle-2 grey--text">child: </span> <span class="font-weight-bold primary--text"><a href="/beam/item/20200207" class="">20200207</a></span>
                (beam)
              </div>
            </div>
          </div>
          <div class="py-2 col col-12">
            <div class="row">
              <div class="py-0 col-md-4 order-md-0 col-12 order-2 align-self-end"><span class="grey--text" style="font-size: 65%;">Data ID: </span></div>
              <div class="py-0 col-md-8 col-12 order-1">
                <div class="caption grey--text">Note</div>
                <div>
                  <ul>
                    <li>This is a dummy test with a lat map</li>
                    <li>A beam depends on this map</li>
                  </ul>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div><button type="button" class="v-btn v-btn--absolute v-btn--flat v-btn--icon v-btn--round theme--light v-size--default" style="top: -15px; right: -10px;"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-nut theme--light grey--text text--lighten-1" style="font-size: 12px;"></i></span></button>
    <div class="v-menu">
      <!---->
    </div>
  </div>
</div>
`;
